### ISSUE TITLE
Tracking Issue for `box_into_inner`

### ISSUE THREAD

**Issue by crlf0710** [rust-lang/rust#80437]
Feature gate: `#![feature(box_into_inner)]`

This is a tracking issue for consuming a `Box` and returning its wrapped value.

The language actually supports `*x` as a special case. However it can't be generalized to other types easily. It also doesn't consume the `Box` value immediately if `T` is `Copy`, which is quite a footgun.

An explicit version is added here to make the API more consistent with `Pin`, `Cell`, `RefCell` etc.

### Public API

This adds the following method to `Box<T, A>`:
```rust
pub fn into_inner(self) -> T;
```

### Steps / History

- [ ] Implementation: ...
- [ ] Stabilization PR

### Unresolved Questions

- None yet.

**Comment by fosskers** [comment]
How come it's a function and not a method?

**Comment by crlf0710** [comment]
@fosskers https://rust-lang.github.io/api-guidelines/predictability.html#smart-pointers-do-not-add-inherent-methods-c-smart-ptr

**Comment by conradludgate** [comment]
When looking at how to turn a box back into a stack owned value, I find this function, but in my case the `*x` method works perfectly fine. It's documented in the box module, but I feel like it should be doubly documented here as a simpler alternative.

**Comment by TehPers** [comment]
Somewhat minor, but I like that `Box::into_inner` can be used in method chains without requiring a closure. `.map(|x| *x)` is still shorter than `.map(Box::into_inner)`, but the latter more clearly describes what you're doing in my opinion. It can also help the compiler infer the type in the iterator/`Option<T>`/whatever at that point in the method chain for what it's worth. Otherwise you'd have to write `|x: Box<_>| *x` for the same effect.

**Comment by BurntSushi** [comment]
`Box::into_inner` appears to have been around for a while, and I don't see any obvious issues with it? What do others think?

@rfcbot merge

**Comment by m-ou-se** [comment]
Untagging lang and restarting the FCP.

@rfcbot cancel

**Comment by m-ou-se** [comment]
@rfcbot merge

**Comment by ZhangHanDong** [comment]
@dtolnay @crlf0710 

I've improved the documentation for `Box::into_inner`, help review it：[Improve documentation for `Box::into_inner` ](https://github.com/rust-lang/rust/pull/98523)

**Comment by Aandreba** [comment]
> which is not a thing that we want people to use this function for, right? We would prefer `assert_eq!(*c, 5)` in that situation still, right?

Why would we prefer * to ```Box::into_inner```?

**Comment by scottmcm** [comment]
> Most (all?) other operators have a functional form that they can be called with.

To me that's not a justification for `Box::into_inner`, though.  It's just another manifestation of the long-wished-for (https://github.com/rust-lang/rfcs/pull/178) `DerefMove`, as a way to remove this special-❄ behaviour from `Box`.

Because if this *was* a general operator that could be used by other types, then yes, it too would have a functional form.

**Comment by dtolnay** [comment]
> Most (all?) other operators have a functional form that they can be called with. \[…\] `&*x` and `&mut *x` being equivalent to `x.deref()`/`x.deref_mut()`, `!x` being equivalent to `x.not()`, etc.

I think this as a justification for `Box::into_inner` misses the point of why those forms exist in the other cases. They exist because we need people to be able to **define** the behavior of deref or logical negation for their type. A trait impl is the language's obvious way to define behavior for a type, to the point where it would be silly to build defining behavior of operators on some other bespoke mechanism instead of trait impls.

Then since they are trait impls, with trait methods, those methods are callable as a side effect. The reason for `.deref()` or `.not()` to be callable is because that's just how trait impls work, and defining behavior of operators is done via trait impls. Not because we felt the need to give people some method to call whenever they don't feel like using an operator.

As a justification it should be clear how none of this applies to `Box::into_inner`.

**Comment by dtolnay** [comment]
> having a function to perform the unboxing operation is convenient for functional code/method chains. `x.iter().map(|x| *x)` works, but `x.iter().map(Box::into_inner)` is more expressive in my opinion

I think this is more compelling than the situation currently described in #98523. Getting a good example along these lines into the documentation would resolve https://github.com/rust-lang/rust/issues/80437#issuecomment-1150255531.

**Comment by Aandreba** [comment]
> Could someone help out with a more compelling justification for this function to exist?

Isn't `Box::into_inner` more readable than dereferencing? I think `into_inner` indicates the value will be moved instead of copied quicker.

**Comment by ZhangHanDong** [comment]
How about this sample code? 

```rust
#![feature(box_into_inner)]
#[allow(unused_mut)]

fn main() {
    let vec = vec![Box::new(1.0), Box::new(2.0), Box::new(3.0)];
    // `v` will be moved, `i` will be moved
    let v : Vec<_> = vec.into_iter().map(|i| {
        let v = Box::into_inner(i) + 1.0; 
        // i;  // Error :use of moved value: `i`
        v
    }).collect();
    assert_eq!([2.0, 3.0, 4.0], &v[..]);

    let vec = vec![Box::new(1.0), Box::new(2.0), Box::new(3.0)];
    // `v` will be moved, `i` will be copied
    let v : Vec<_> = vec.into_iter().map(|i| *i + 1.0).collect();
    assert_eq!([2.0, 3.0, 4.0], &v[..]);

    let vec = vec![Box::new(1.0), Box::new(2.0), Box::new(3.0)];
    // Error: cannot move out of `*i` which is behind a shared reference
    // let v : Vec<_> = vec.iter().map(|i| Box::into_inner(*i) + 1.0).collect();

    let vec = vec![Box::new(1.0), Box::new(2.0), Box::new(3.0)];
    // `v` will be moved, `i` will be copied
    let v : Vec<_> = vec.iter().map(|i| **i + 1.0 ).collect();
    assert_eq!([2.0, 3.0, 4.0], &v[..]);
}

```

[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=8747318c1e4e5b5881f656b9163cad4e)

**Comment by npmccallum** [comment]
@dtolnay I have a real-world use case. I have a trait `Foo`. This trait contains a method which consumes `self`. Most of the time this trait is implemented directly on types. But occasionally, I need a trait object for `Foo`. Therefore, I want to provide a blanket implementation of the trait for a box of a type which likewise implements the trait. For example:

```rust
#![feature(box_into_inner)]

pub trait Foo {
    // ... other stuff ...

    fn finish(self);
}

impl<T: Foo> Foo for Box<T> {
    // ... other stuff ...

    fn finish(self) {
        Box::into_inner(self).finish()
    }
}
```

**Comment by scottmcm** [comment]
@npmccallum It's unclear to me why `(*self).finish()` isn't fine there?

That'd be more consistent with how one writes `&`/`&mut` methods in such cases too, as can be seen on the standard library's implementation of `Iterator` for boxed iterators, for example: https://doc.rust-lang.org/1.62.0/src/alloc/boxed.rs.html#1803-1817

**Comment by Aandreba** [comment]
> It's unclear to me why `(*self).finish()` isn't fine there?

`Box::into_inner(self).finish()` better signals the intention to move the value from the box, instead of copying the value from it.
Also, in my opinion, it looks beter.

**Comment by fundon** [comment]
Here's a very practical use case for a routing system. Likes [actix-service](https://docs.rs/actix-service/latest/actix_service/) and [tow-service](https://docs.rs/tower-service/latest/tower_service/).

Whenever you want to add a new middleware, you need to boxed operation, if you have into_inner method, you only need to boxed once is enough.

**Comment by dtolnay** [comment]
I don't understand how `Box::into_inner` helps reduce nesting. Could you please say more?

**Comment by fundon** [comment]
Here is an example using [actix-web](https://github.com/actix/actix-web): 

https://github.com/actix/actix-web/blob/master/actix-web/src/route.rs#L40-L63
```rust
    /// Registers a route middleware.
    ///
    /// `mw` is a middleware component (type), that can modify the requests and responses handled by
    /// this `Route`.
    ///
    /// See [`App::wrap`](crate::App::wrap) for more details.
    #[doc(alias = "middleware")]
    #[doc(alias = "use")] // nodejs terminology
    pub fn wrap<M, B>(self, mw: M) -> Route
    where
        M: Transform<
                BoxService<ServiceRequest, ServiceResponse, Error>,
                ServiceRequest,
                Response = ServiceResponse<B>,
                Error = Error,
                InitError = (),
            > + 'static,
        B: MessageBody + 'static,
    {
        Route {
            service: boxed::factory(apply(Compat::new(mw), self.service)),
            guards: self.guards,
        }
    }
 ```
 
 https://github.com/actix/actix-net/blob/master/actix-service/src/boxed.rs#L66-L80
 
 ```rust
 /// Wraps a service factory that returns service trait objects.
pub fn factory<SF, Req>(
    factory: SF,
) -> BoxServiceFactory<SF::Config, Req, SF::Response, SF::Error, SF::InitError>
where
    SF: ServiceFactory<Req> + 'static,
    Req: 'static,
    SF::Response: 'static,
    SF::Service: 'static,
    SF::Future: 'static,
    SF::Error: 'static,
    SF::InitError: 'static,
{
    BoxServiceFactory(Box::new(FactoryWrapper(factory)))
}
 ```
 
 `self.service` is boxed, if add a new middleware to it, need to wrap it again use `Box`. 
 When adding a lot of middleware, it wraps many `Box` layers.
 
 If we provide the `Box::into_inner` method, we can unpack it first and then add it to the final box.

**Comment by dtolnay** [comment]
But that could be written easily and concisely (with less nesting of parentheses) as:

```rust
boxed::factory(apply(Compat::new(mw), *self.service))
```

so this does not address https://github.com/rust-lang/rust/issues/80437#issuecomment-1150255531.

**Comment by Aandreba** [comment]
Perhaps this would be a good example for [#80437 (comment)](https://github.com/rust-lang/rust/issues/80437#issuecomment-1150255531).

This compiles
```rust
let a = Box::new(5);
let b = *a;
let c = *a;
```

This doesn't
```rust
#![feature(box_into_inner)]

let a = Box::new(5);
let b = Box::into_inner(a); // move occurs because `a` has type `Box<i32>`, which does not implement the `Copy` trait
let c = *a; // error, use of moved value: `a`
```

**Comment by dtolnay** [comment]
That looks like what's already in #98523, but more abstract and less fleshed out / less compelling. There is no rationale for why someone would want the first code block to not compile. It looks totally nonproblematic.

**Comment by crepererum** [comment]
> Could someone help out with a more compelling justification for this function to exist?

One application could be the interplay with `Any`. If I'm not mistaken, you can move `Any` types only in boxes, i.e. `Box<dyn Any>`. Now if you downcast to a concrete type you end with `Box<MyType>` but there is no way to remove the box in the end.

**Comment by dtolnay** [comment]
Not correct, `*` removes the Box. There is no need at all for into_inner in that situation.

**Comment by Mathspy** [comment]
The only reason I am (and probably a bunch of others) are subscribed to this issue is because when I was unfamiliar with `Box` it didn't instantly occur to me that I can just `*boxed_value` to get the value, so I looked up `Box::into_inner`, found this method and was confused it was not stable, opened the issue and subscribed and then realized from the comments "Oh I can just use `*`, I am silly".
I know there's no precedence for this but would it be possible to leave this unstable forever just to have its documentation point people towards using `*`?

**Comment by crepererum** [comment]
One argument could be that `into_inner` is easier to find than the special "`DerefMove`" semantic of `Box`. There could be a clippy lint on top that suggest replacing `into_inner` with `*`. It's a somewhat weak argument though to add stdlib features just for learners.

**Comment by alamb** [comment]
> Could someone help out with a more compelling justification for this function to exist?

In my unlearned opinion the reason for the function to exist is because newcomers to rust expect it to exist and thus are confused when it doesn't.

Expert rust programmers will use `*` and clippy could guide newcomers to the correct usage of `*` as @crepererum  suggests in https://github.com/rust-lang/rust/issues/80437#issuecomment-1218133185

**Comment by dtolnay** [comment]
I am skeptical of a claim that newbies expect it to exist without looking at the Box docs, especially if that's in the form of correctly guessing Self::assocfn syntax over .method() syntax, which is idiosyncratic to Box. None of this strikes me as a thing newbies guess.

If instead, newbies expect it to exist because they look at the Box docs and it exists, then that that is relatively easy to fix by making the function not exist, and then improving the docs to explain how to use Box in the intended way.

**Comment by Mathspy** [comment]
A bit of an anecdotal evidence here: the only reason I ever subscribed to this issue was because I rarely work with Box and one time when I was doing so I needed to “unbox” and so my initial thought was to look for an into_inner method similar to the ones on Mutex and other types.
I found the method and was surprised it wasn’t stabilized, expected it to be a 1.0.0 stable, opened the issue and subscribed and while reading I realized that `*` on Box is basically `into_inner` and that I can just use that
There are probably a non-zero amount of people who might go through something similar, I don’t think there’s any precedence for this but maybe it’d be worth keeping the method unstable forever and changing its documentation to point to `*`?

Either ways I don’t mind whichever direction but wanted to leave my experience here in case it helps make a decision

**Comment by conradludgate** [comment]
Similar experience here, but I'm with @dtolnay on this one. I was confused by the `into_inner` method and I assumed it was the only way. If it hadn't autocompleted in my editor I would have most likely searched the box docs more thoroughly.

The only argument I have for `into_inner` is slowly moving us away from magic box-deref into some other language feature, but I don't think I've seen much proposal related to that here (I think DerefMove is a dead idea atm 😞)

**Comment by alamb** [comment]
To be clear, it is my opinion that adding the following would make rust easier for newcomers. 

```rust
let value = my_boxed_value.into_inner();
```

I clearly can't speak for all newbies -- only my past self. I read various rust books, thought I understood `&` and `*` and all was good and then I had a `Box` and had to get the value out of it. I was expecting some function to do so and finding none it took a few minutes  go google / get out of the my coding flow, and read about the very special case of `*` when applied to Box. 

So yes I think the "RTFM" response is reasonable (and a common one from subject matter experts), but it is my opinion that having a `into_inner()` method and then a clippy lint would improve the experience and remove one (small) rough edge in learning the Rust language 🤷

**Comment by conradludgate** [comment]
> To be clear, it is my opinion that adding the following would make rust easier for newcomers.
> 
> ```rust
> let value = my_boxed_value.into_inner();
> ```

To be extra clear. This tracking issue is for:

```rust
let value = Box::into_inner(my_boxed_value);
```

Making it a method is a breaking change

**Comment by dtolnay** [comment]
That's right, speaking for the library API team, a `.into_inner()` method is not a thing we would consider adding. Ease of learning is relevant but that is not a solution to it.

**Comment by alamb** [comment]
> Ease of learning is relevant but that is not a solution to it.

Well, not your preferred solution, perhaps

**Comment by plafer** [comment]
My experience of learning the special dereference case for `Box` echoes the experience of others in this thread, and certainly confused me.

IMO special cases make the language harder to understand. Take someone learning the language, playing with moves, derefs, deref coercion, the dot operator, etc. and forming a model of how these features work. When they get to the dereference behavior of `Box`, they'll either be surprised and revisit their model, or blindly confirm an inaccurate model (e.g. that one can move a value of a `Deref` type with the `*` operator). In other words, the special case for `Box` doesn't just affect the learnability of `Box`, but of dereferencing in general.

Although this RFC doesn't remove the dereference special case of `Box`, IMHO it's a step in the right direction towards reducing the complexity of Rust.

**Comment by scottmcm** [comment]
> Although this RFC doesn't remove the dereference special case of `Box`, IMHO it's a step in the right direction towards reducing the complexity of Rust.

I don't think this method actually does that, because it doesn't remove the special behaviour from `Box`, and thus one still needs to know it when reading code.

The long-term goal for reducing the complexity is to make `Box` not special by finding a way for *other* types to get this behaviour from `*` too.  (This is often called `DerefMove`, though the details have not been worked out.)

**Comment by Aandreba** [comment]
I really don't know what the problem is, the case has been clearly made.

- Makes regular code more readable
- Signals the intent to move the value from the box instead of copying it
- It avoids the non-standard use of `*` on boxes
- It makes functional code more readable (`Box::into_inner` v. `|x| *x`)

If these are not considered valid points for the stabilization of `into_inner`, there's honestly not much else to discuss. 

I personally use it every time I need to take a value from a box, because this is the way it **should** be done, the same way you should cast with `Into`/`From`/`TryInto`/`TryFrom` instead of `as` (usually)

**Comment by Aandreba** [comment]
> Huh? You're saying pointer dereference is non-standard?

Regular pointer dereferencing (where you get a copy of the value pointed too) is very much standard.
Boxed pointer dereferencing (where you may get a copy of the value or take ownership of it) is very much non-standard.

**Comment by azriel91** [comment]
> Is there anything Box::into_inner gives you that * does not?

It can provide clarity / intent. My mind more readily maps:

* `into_inner` as "getting the data from inside this thing", and
* `*x` first maps to "dereference this thing, which may be a `Deref::Target`", before "get the thing"

**Comment by Aandreba** [comment]
> Is there anything `Box::into_inner` gives you that `*` does not?

What I mentioned on [my previous list](https://github.com/rust-lang/rust/issues/80437#issuecomment-1238519736).

Also, not all additions to the standard library should be new features, having convenince function in it is quite usefull. If we didn't have convenience functions, 90% of the methods for `Option` and `Result` wouldn't exist (you can do basically all of them with regular pattern expressions) .

**Comment by BurntSushi** [comment]
What I meant is whether it gives any new powers. It sounds like the answer to that is no.

> Also, not all additions to the standard library should be new features

Well... Yes? But it's also important to know whether an addition is about a new feature or whether it's about a convenience.

When I proposed stabilizing this, I believe I was under the impression that it was a new feature. That is, that it lets you do something you couldn't do before. But it looks like that's not true, and I can't remember precisely what I had in mind.

I'm more inclined to agree with dtolnay here. I'm not really convinced this routine improves clarity or intent enough to be worth it.

**Comment by Aandreba** [comment]
> I'm not really convinced this routine improves clarity or intent enough to be worth 

It just unacceptable the way dereferencing boxes works, it breaks standard by making an exception for a specific type, and that type only. If I have a `Box<i32>`, I'll never know if the value has been copied or taken, and even though I know that under the hood this matters very little, not knowing the exact way your code works is a recipe for disaster and a rabbit hole we should not go down.

What's done is already done, we can't undo that decision now (unless we want to break backwards compatibility), but we can very much offer a **correct** alternative.

**Comment by conradludgate** [comment]
If the goal is to work on consistency, I would much rather we focus efforts on a DerevMove trait and not this method. As it stands, the implementation of this function is just `*boxed` which anyone can write.

```rust
pub fn box_into_inner<T>(boxed: Box<T>) -> T {
    *boxed
}
```

Not as nice as being in std, but it's not much overhead if you really want a function that imitates this behaviour.

**Comment by Aandreba** [comment]
> As it stands, the implementation of this function is just `*boxed` which anyone can write.

A big part of the standard library are functions that could had been written manually (see `Option` and `Result` methods). Also, what downside could stabilizing `into_inner` have? As far as I see, none, and similar functions exist for other types (like all the cells).

**Comment by BurntSushi** [comment]
> Also, what downside could stabilizing `into_inner` have?

Decision paralysis. There is already a way to utter `Box::into_inner` in Rust and that's `*`.

> it breaks standard by making an exception for a specific type

I don't think anyone is happy about `Box` being "special." So I mean, like, yeah, we're all on the same page there. That doesn't mean we should add `Box::into_inner`.

> but we can very much offer a **correct** alternative.

Are you saying `*x` where `x` is a `Box<T>` is incorrect? I'm confused. I thought we just established that this method was about clarity/intent, but now it seems like you're saying it's about correctness? Can you provide some real code that needs `Box::into_inner` to be correct?

**Comment by Aandreba** [comment]
> Are you saying `*x` where `x` is a `Box<T>` is incorrect?

Not in the sense that it produces incorrect code.

**Comment by Aandreba** [comment]
> I would much rather we focus efforts on a DerevMove trait and not this method

Why would `DerefMove` be better than `into_inner`?

**Comment by Aandreba** [comment]
> Decision paralysis. There is already a way to utter `Box::into_inner` in Rust and that's `*`

The same argument could be made for number conversion. Do I use the `Into`/`From`/`TryInto`/`TryFrom` traits or do I use the `as` keyword?

**Comment by npmccallum** [comment]
@Aandreba `From` and `as` do different things. `From` is lossless. `as` is potentially lossy.

**Comment by BurntSushi** [comment]
> The same argument could be made for number conversion. Do I use the Into/From/TryInto/TryFrom traits or do I use the as keyword?

I wasn't making an argument though. You asked for a _cost_. I gave you one. Do we have other "decision paralysis" costs in the standard library? Yes! Absolutely! There's a difference between the questions "what are the downsides" and "should we do this." You said, "Also, what downside could stabilizing into_inner have?" Which is clearly an instance of the former, not the latter.

Otherwise, this conversation is honestly getting pretty frustrating to participate in, so I'm going to bow out until there's something more significant to respond to. I'll say that I am not _against_ this API landing, but I am not strongly in favor of it either.

**Comment by Aandreba** [comment]
> You said, "Also, what downside could stabilizing into_inner have?" Which is clearly an instance of the former, not the latter.

That's true. The point I'm trying to make is that I don't see how the decision paralysis cost outweighs the benefit in code readability. And it's also a very small decision for which you'll have a goto answer by the second time you encounter it.

**Comment by TehPers** [comment]
I also think that this function is better served in a more generic `DerefMove` trait, which would allow for a more widely usable version of `into_inner`. Because of that, I'm not sure that `Box::into_inner` is any better than, for example, a `DerefMove::deref_move` function. The biggest reason I want this is for convenience during function chains (in particular, `.map(Box::into_inner)`), but that is also solved by `DerefMove`. I would be more in favor of that feature stabilizing, personally.

**Comment by Aandreba** [comment]
I see `DerefMove` as a bad idea. To me it looks like "ups, we added a bad exception for a type, let's solve it by making it possible to apply the same exceptions to other types". I think `into_inner` is a better alternative.

**Comment by TehPers** [comment]
I'm curious what makes the move-on-derefence behavior bad for `Box<T>`? To me, it seems like a very useful convenience, given how many functions take `self` and how often you work with boxed values. On the other hand, generalizing it seems like it would open the possibility for alternate pointer types to take advantage of that behavior. That being said, discussion around `DerefMove`  specifically is probably better suited in the `DerefMove`/`IndexMove`/`IndexSet` issue.

On the topic of this issue, calling `Box::into_inner` every time I needed to move out of a `Box<T>` would be quite a bit of syntactic overhead every time I needed to pass a `self`. While I would like an `into_inner`-like function, I wouldn't like being forced to use it everytime, especially when `Deref` and `DerefMut` already allow you to auto-deref pointer types to call `&self`/`&mut self` methods on types behind pointers:

```rs
let mut s = Box::new(String::new());
s.push('a'); // The boxed value is automatically dereferenced here with DerefMut's implementation
```

**Comment by Aandreba** [comment]
> discussion around DerefMove specifically is probably better suited in the DerefMove/IndexMove/IndexSet issue.

Agreed. But the point was being made that with `DerefMove::deref_move`, `Box::into_inner` would be unnecessary (which I agree). My point was "I don't think DerefMove is a good idea, so I hope it never comes to bee. If it doesn't, we'll still have the same problem as now"

> calling Box::into_inner every time I needed to move out of a Box<T> would be quite a bit of syntactic overhead every time I needed to pass a self.

Well, I don't think anybody is arguing we should remove the special behavior `*` has on boxes. The Rust team has made a clear commitment to to backwards compatibility, and that would break it.

> I wouldn't like being forced to use it everytime, especially when Deref and DerefMut already allow you to auto-deref pointer types to call &self/&mut self methods on types behind pointers

The problem discussed here is not about taking references of the values inside a box. The `Deref`/`DerefMut` API is a well specified and designed one, and yes, 99% of the time you'll work with the value still boxed. 

But when time comes to take ownership of the boxed value in the stack, I (and many others) think `Box::into_inner` is preferable to `*`.

**Comment by conradludgate** [comment]
> Well, I don't think anybody is arguing we should remove the special behavior `*` has on boxes. The Rust team has made a clear commitment to to backwards compatibility, and that would break it.

This is just syntax. Syntax can be changed in editions, and I have heard people genuinely suggesting we should remove the `*` special behaviour.

> But when time comes to take ownership of the boxed value in the stack, I (and many others) think `Box::into_inner` is preferable to `*`.

Logically, `*` makes sense for box. We use `*` to get to the referee. In the case of `&T`, either we need to copy `T` or we need to immediately `&` it again otherwise this is not valid. However, since we have `Box<T>` as an 'owned pointer', it does make sense that this is not required.

My personal issue is that this behaviour is currently special to Box. I don't mind language features, but we typically have a trait that encodes those behaviours: 
* `for` is a wrapper around `Iterator`
* `+` is `Add`
* `.await` is `Future`
* `*` is already overloaded with Deref/DerefMut depending on the context of needing a `&T/&mut T` respectively. I think adding a third trait that we can point to, even if we seal it just for box to implement, is a good thing. Then we can do `Box::deref_move` instead of `Box::into_inner` for the same effect.

**Comment by LunarLambda** [comment]
Has there been any further discussion about this? I (and people I've taught Rust to) have gotten quite surprised more than once at `*box` being the way to move out of a box.

**Comment by EndilWayfare** [comment]
Just to add my two cents, I _just today_ found out about how "unbox" has to be spelled `*x`, despite working pretty heavily in Rust for almost 5 years. 

For some reason, in all my use cases, I've never had a reason to reach for a `Box` in situations where ownership transferred. I've used other smart-pointer/cell types extensively, because most of the time when I need a heap allocation that isn't a standard collection or a foreign crate import, the _reason_ I'm doing so is better served by an `Rc` or `Arc`, with or without a `RefCell`/`Mutex`. I've even had reasons to _manually, explicitly use `Pin`_. I _knew_ that `Box` has some special-casing in `rustc`.

Given that prior context, my immediate thought was "oh, obviously `Box::into_inner`." But it wasn't. And I had to look it up.

I wouldn't consider myself a Rust "expert". But, my point is, I'm certainly not a noob at this point, either, and I found this behavior _surprising_. In fact, _surprisingly_ surprising. One of my (many) favorite parts about Rust is the consistency, coherence, and wide-adoption of naming convention patterns. They're short, sweet, descriptive, readable, say-able, and practically self-documentation in many cases. That's probably the main reason why I expected `Box::into_inner` to be the idiom, and expected _pretty confidently_ at that.

Still anecdotal/personal-preference, but I figured it might be relevant that the issue isn't unique or specific to newcomers.

**Comment by SylvKT** [comment]
> I wouldn't consider myself a Rust "expert". But, my point is, I'm certainly not a noob at this point, either, and I found this behavior _surprising_. In fact, _surprisingly_ surprising. One of my (many) favorite parts about Rust is the consistency, coherence, and wide-adoption of naming convention patterns. They're short, sweet, descriptive, readable, say-able, and practically self-documentation in many cases. That's probably the main reason why I expected `Box::into_inner` to be the idiom, and expected _pretty confidently_ at that.

I agree with all of what you said. I never thought unboxing a `Box` would be `*boxed`, so I think this should be—*at the very least*—locked behind an `#[allow()]` in the next Rust edition, if not entirely removed.

**Comment by ais523** [comment]
Currently, the `*boxed` method of converting a box into its inner value is underdocumented and I learned about it only today, from this bug report (which I found by searching `Box::into_inner()` in the documentation), despite having been using Rust for over 6 years. It is not listed in the documentation for `std::boxed` or `std::boxed::Boxed`. It is not listed in the section of the Rust Reference that covers a `DereferenceExpression`. "The Rust Programming Language" not only does not mention that a `Box` can be used like that, it actively denies it (there is an entire section on how calling `*` on a box uses the `Deref` trait and how you need to implement the `Deref` trait yourself to get the same behaviour on your own smart pointers, and a statement that "Boxes provide only the indirection and heap allocation; they don’t have any other special capabilities, like those we’ll see with the other smart pointer types.").

If the intended way to convert a box into its inner value is `*boxed`, then this urgently needs to be documented somewhere people will find it – it is possible that it is documented somewhere, but I didn't find it documented anywhere other than this bug report. It would make a lot of sense to at least, in the documentation of the unstable `box_into_inner`, to say "You can also convert a box into its inner value by doing `*boxed`" and giving an example of that – the documentation of `Box` is an obvious place to look for someone who's trying to do that in their code, and `into_inner` is the obvious method to look at. I got here because, finding no stable way to do the conversion, I decided to check the bug report to see how close it was to stabilisation.

Note that updating the documentation for `into_inner` to list the currently stable way to do things is something that doesn't require it to actually be stabilised to have benefits; even when people use stable versions of Rust, they use documentation that lists the unstable methods, so that would be a good way to have discoverability for the feature in question (and there is no reason to not make it discoverable even if planning to remove it in an edition – you would need to maintain backwards compatibility for code that's already using it anyway). As such, it could be done regardless of the conclusions in this issue as to which of the two syntaxes is preferable.

**Comment by ais523** [comment]
@BurntSushi: The example given actually copies a value out of the box back to the stack, using the `Box as Deref` trait implementation, because `u8: Copy`. For example, this works (and the first two lines are the given example):
```rust
let boxed: Box<u8> = Box::new(5);
let val: u8 = *boxed; 
println!("{}", val);
println!("{}", boxed);
```
[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=14f3213025195bfa32e8543e9b314516)

If the `Box` were consumed, this wouldn't compile (because a `Box<u8>` is not `Copy` even though a `u8` is).

As such, the example isn't very instructive to people who are trying to learn how `Box` works because they just assume it's a demonstration of `Box` supporting `Deref`, something which is comprehensively discussed in the documentation. (Yes, the introduction to the example says "moved" – but that's inconsistent with the actual example, which is a demonstration of `Deref` and not `DerefMove`.) In particular, it doesn't make it clear what would happen with a `Box` that boxes a non-`Copy` type, and as such, isn't teaching people about the special case that moves the contents.

**Comment by ais523** [comment]
I wonder if we're looking at different documentation? I'm looking at https://doc.rust-lang.org/std/boxed/index.html in which the example in question is
```rust
let boxed: Box<u8> = Box::new(5);
let val: u8 = *boxed;
```

It is possible that this has already been fixed recently, and the fix simply hasn't reached the public documentation yet.

(I also can't submit pull requests via Github because doing so would force my account to use two-factor authentication, and I don't have a smartphone or other similar device to use as the second factor.)

**Comment by BurntSushi** [comment]
Yes. That's the same example I'm looking at. You can replace `u8` with a non-`Copy` type and it still works.

**Comment by ais523** [comment]
So the problem is that the example doesn't convey the information that it would still work after replacing `u8` with a non-`Copy` type. Someone who had read most of the relevant documentation (e.g. The Rust Programming Language and the Rust Reference) would be under the impression that the *reason* the example worked was that `*` is calling the `Deref` implementation on `Box` to obtain an `&u8`, and that it is possible to copy out of an `&u8`. It still works if you replace the `Box` with something that doesn't have the specail case:

```rust
use std::rc::Rc;
let boxed: Rc<u8> = Rc::new(5);
let val: u8 = *boxed;
````
[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=d927535739ce991a99765d8dab91cdd2)

As such, the example fails to inform readers that the special case exists – because it would work even without it (and in fact doesn't make use of it).

**Comment by BurntSushi** [comment]
OK, I'm not sure I really agree with that. But I don't really know how to engage with you further without looking at a concrete set of edits to the status quo. I don't know how to resolve your issue with PRs. If you can't create one, perhaps you can find someone to do it on your behalf.

I also think this is meandering into off-topic territory for this specific issue. I might suggest creating a new one because "documentation for `Box` is not as good as it could be" and "should we stabilize `Box::into_inner`" are, while related, two distinct things.

**Comment by dtolnay** [comment]
We discussed this tracking issue in today's standard library API team meeting. The thing it's stuck on is still a PR for https://github.com/rust-lang/rust/issues/80437#issuecomment-1178175428. Some of the comments following that one have some additional suggestions.

**Comment by Bwallker** [comment]
```rust

#[cfg(some_condition)]
/// Some custom box type that we're using. Maybe we're using a custom allocator or maybe we don't want noalias.
struct CustomBox<T>(/* TODO: Implement this. */ PhantomData<T>);

#[cfg(some_condition)]
impl<T> CustomBox<T> {
    fn into_inner(this: Self<T>) -> T {
        todo!();
    }
}

#[cfg(some_condition)]
type MyBox<T> = CustomBox<T>;

#[cfg(not(some_condition))]
type MyBox<T> = Box<T>;

fn foo<T>(t: MyBox<T>) {
    // Get inner value.
    let inner = MyBox::into_inner(t);
}

```

One use case I haven't seen mentioned is when you have a type alias that may or may not point to std::Box<T> and you want to get the inner T. This could also be solved by stabilizing MoveDeref, but MoveDeref has been stalled for several years as far as I'm aware, so I think stabilizing this is more possible/actionable in the short to medium term.

**Comment by ActuallyHappening** [comment]
Haven't read the whole conversation, but this method was exactly what I was looking for. My mental model tells me that going from an owned pointer to the inner should happen, so looking for this method I was surprised that it wasn't stabilized yet. But `*state` was good enough for the moment :)

**Comment by ActuallyHappening** [comment]
Would still like this to be stabalized

**Comment by theemathas** [comment]
A notable thing about this is: `let x = *my_box;` does *not* immediately deallocate the heap memory. The heap memory only gets deallocated when `my_box` (which now contains the "husk" of a box) goes out of scope.

In contrast, `let x = Box::into_inner(my_box);` immediately deallocates the box, since the entire box is consumed. Emulating this without a function call requires writing the obscure incantation of `let x = *{my_box};`

Is this a strong reason to stabilize this function?

**Comment by tgross35** [comment]
With allocator stabilization around the corner, perhaps we should consider making this `-> (T, A)`?
